text/0000-edition-2021.md

```diff
@@ -0,0 +1,238 @@
+
+Some examples of changes editions cannot make:
+
+* Loosening the coherence rules in just one edition.
+    * Coherence is a "zero sum" global property that states that there are no two impls that apply to the same types. This property must hold across all crates, regardless of edition, so we can't tweak the rules in just crates in the 2021 Edition to permit impls that would be disallowed in 2018 crates. If we did so, then crates in the 2018 Edition could add the impl under the old rules, and another crate in the 2021 Edition could add the same impl per the new rules, resulting in an error.
+
+[RFC 2229]: https://github.com/rust-lang/rfcs/pull/2229
+
+## Access to the 2021 edition on nightly
+
+The nightly compiler will make a 2021 edition available in an unstable form. This will allow us to build and test the migrations for various features that will be part of the 2021 edition. 
+
+**Warning:** It is not recommended for crates, even nightly-only crates, to adopt the 2021 Edition too early. Like any unstable feature, the future editions can change without notice and crates are expected to keep up. Furthermore, the automatic migrations are designed to be used exactly once; if a crate C is migrated from Edition X to Edition Y, and then new features or migrations are added to Edition Y, the crate C may not be able to access those new migrations and the changes may have to be done manually.
+
+## Advertising and documenting the edition
+
+The expectation is that the 2021 Edition will not be marketed as an "all encompassing event" in the same way as the 2018 Edition.
+
+Producing an edition requires producing the following key artifacts. This list is not meant to be exhaustive; as part of executing on the 2021 Edition, we will be producing more detailed guidelines for how the process works that can be used as a template for future editions if desired.
+
+* "What is an edition" page on the main Rust web site
+* Edition migration guide
+* Edition status tracking page, likely part of a general tracking page as envisioned by [RFC 3037](https://github.com/rust-lang/rfcs/pull/3037).
+* Blog post announcing the edition -- this will be the "highlight" part of the standard release blog post
+
+## Metrics for a successful edition
+
+The following metrics are an attempt to quantify some of the goals we have for editions.
+
+### Easy migration
+
+* 90% of crater-seen crates migrate without manual intervention.
+    * This can be measured automatically, subject to the limitations of crater (e.g., linux only).
+* 90% of crates that were migrated to the newer edition could be migrated in less than 1 hour
+    * This includes both rustfix but also manual migration.
+    * This will require a survey or other forms of self-reporting.
+
+### Adoption
+
+* 75% of Rust survey respondents in 2022 indicate that they are using Rust 2021
+    * If we don't see this number, it may not be a problem, but it'd be worth investigating why.
```

Comment:
Note that there's non-negligible number of crates that provide some MSRV guarantees. I personally provide "latest rustc in Debian stable" MSRV in all my first-class crates.

75% in 2022 feels very high. If this makes popular, important crates like Tokio to bump MSRV then God help us as I highly doubt the new edition will make it into Bullseye in 2022. Tokio not supporting Buster is already a huge pain even though the changes are cosmetic and the maintainers completely ignored my offer for a PR.

Reply:
To be quite honest, I pulled this threshold from thin air. In the past though we saw very high adoption. Still, Rust 2018 contained a number of banner features and Rust 2021 does not, so 70% may indeed be very high.

Reply:
Right, I've heard complains that Rust develops too fast (as in new language changes) and I kinda agree, especially it makes lives of people who want to progress more slowly/carefully difficult.

To be more constructive, how about 50% in crates without MSRV policy within two years since release?

Reply:
I think this is more so addressing a wider question of what MSRV policies in the Rust ecosystem are and how the Rust project positions itself toward that audience; currently, I think the 75% goal gives us something reasonable to shoot for. If we find that a significant percentage of rust users respond with "have not migrated, need MSRV X" that will also be helpful. Ultimately in practice the key thing here is that we set out to measure this question, not really more than that.

Reply:
@Mark-Simulacrum if you put it that way, I agree. Thanks for clarification! :)